crates/store/re_chunk/src/slice.rs

```diff
@@ -530,6 +560,172 @@ impl Chunk {
+            heap_size_bytes: _,
+            is_sorted,
+            row_ids,
+            timelines,
+            components,
+        } = self;
+
+        // Safe early out to prevent panics in upstream kernel implementations.
+        if filter.len() != self.num_rows() {
+            return None;
+        }
+
+        if self.is_empty() {
+            return Some(self.clone());
+        }
+
+        let num_filtered = filter.values_iter().filter(|&b| b).count();
+        if num_filtered == 0 {
+            return Some(self.emptied());
+        }
+
+        re_tracing::profile_function!();
+
+        let is_sorted = *is_sorted || num_filtered < 2;
+
+        let mut chunk = Self {
+            id: *id,
+            entity_path: entity_path.clone(),
+            heap_size_bytes: Default::default(),
+            is_sorted,
+            row_ids: crate::util::filter_array(row_ids, filter),
+            timelines: timelines
+                .iter()
+                .map(|(&timeline, time_column)| (timeline, time_column.filtered(filter)))
+                .collect(),
+            components: components
+                .iter()
+                .map(|(&component_name, list_array)| {
+                    let filtered = crate::util::filter_array(list_array, filter);
+                    (component_name, filtered)
+                })
+                .collect(),
+        };
+
+        // We can know for sure whether the resulting chunk is already sorted (see conditional
+        // above), but the reverse is not true.
+        //
+        // Consider e.g. densifying the following chunk on `example.MyPoint`:
+        // ┌──────────────┬───────────────────┬────────────────────────────────────────────┐
+        // │ frame        ┆ example.MyColor   ┆ example.MyPoint                            │
+        // ╞══════════════╪═══════════════════╪════════════════════════════════════════════╡
+        // │ 3            ┆ [4278255873]      ┆ -                                          │
+        // ├╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
+        // │ 1            ┆ -                 ┆ [{x: 1, y: 1}, {x: 2, y: 2}]               │
+        // ├╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
+        // │ 5            ┆ -                 ┆ [{x: 3, y: 3}, {x: 4, y: 4}, {x: 5, y: 5}] │
+        // └──────────────┴───────────────────┴────────────────────────────────────────────┘
+        //
+        // The original chunk is unsorted, but the new filtered one actually ends up being sorted.
+        chunk.is_sorted = is_sorted || chunk.is_sorted_uncached();
+
+        #[cfg(debug_assertions)]
+        #[allow(clippy::unwrap_used)] // debug-only
+        chunk.sanity_check().unwrap();
+
+        Some(chunk)
+    }
+
+    /// Applies a [take] kernel to the [`Chunk`] as a whole.
+    ///
+    /// In release builds, indices are allowed to have null entries (they will be taken as `null`s).
```

Comment:
What are the situations that cause us to query with null indices? Seems like returning a `ChunkResult` here and always making that an error condition would be preferable.

Reply:
We don't, but this is technically part of the public Rust API, so I don't want to punish end users trying to do something that is perfectly valid and apparently well accepted in the broader ecosystem (whether its panics or results, they're both extremely annoying in these filter chains).